h2o-core/src/main/java/water/rapids/ast/prims/operators/AstBinOp.java

```diff
@@ -329,15 +343,47 @@ public void map(Chunk[] chks, NewChunk[] cress) {
           if (clf.vec().isString())
             for (int i = 0; i < clf._len; i++)
               cres.addNum(str_op(clf.atStr(lfstr, i), crt.atStr(rtstr, i)));
-          else
+          else if (categoricalOK() && clf.vec().isCategorical() && crt.vec().isCategorical()) {
+            // The vec with longer domain is iterated over due to categorical mapping
+            if (clf.vec().domain().length > crt.vec().domain().length) {
+              for (int i = 0; i < clf._len; i++) {
+                cres.addNum(op(clf.atd(i), alignedCategoricals[c][(int) crt.atd(i)]));
+              }
+            } else {
+              for (int i = 0; i < clf._len; i++) {
+                cres.addNum(op(alignedCategoricals[c][(int) clf.atd(i)], crt.atd(i)));
+              }
+            }
+          } else {
             for (int i = 0; i < clf._len; i++)
               cres.addNum(op(clf.atd(i), crt.atd(i)));
+          }
         }
       }
     }.doAll(lf.numCols(), Vec.T_NUM, new Frame(lf).add(rt)).outputFrame(lf._names, null);
     return cleanCategorical(lf, res); // Cleanup categorical misuse
   }
 
+  /**
+   * Produces a mapping array with indexes of the smaller pointing to the larger domain.
+   *
+   * @param longerDomain Domain to originally map from
+   * @param shorterDomain Domain to originally map to
+   * @return Cross-domain mapping as an array of primitive integers
+   */
+  private int[] alignCategoricals(String[] longerDomain, String[] shorterDomain) {
+    Arrays.sort(longerDomain);
```

Comment:
This is being called this way:

    alignCategoricals(lf.vec(c).domain(), rt.vec(c).domain())

Such call will mutate the domain of the Vec. This is undesirable because the original domain mapping will get lost. Vec will typically (99%) have domain already sorted but there is no guarantee (eg. a rename operation will produce an unordered domain).

Reply:
The alignCategoricals method now creates a copy of the longerDomain before sorting it and performing binary search internally. It will never mutate any String[] argument - ensured. Sorting the shorter one is avoided of course !